### PR TITLE
Fix `map_entry` false positive

### LINKED ISSUE #4219: false positive for map_entry when two different maps used
Clippy wrongly triggers the map_entry lint in the following case (note the map used with `contains_key` is different from the one used with `insert`):

```rust
    let a: HashMap<String, String> = HashMap::new();
    let mut b: HashMap<String, String> = HashMap::new();

    let key = "foo".to_string();
    
    if !a.contains_key(&key) {
        b.insert(key, "baz".to_string());
    }
```

Link to the playground (click Tools -> Clippy): https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b230b4300359f8ddaa4e2610d2e0dd1c

```
> cargo clippy -V
clippy 0.0.212 (265318d 2019-05-17)
```

### PR DESCRIPTION
**Pull request by JohnTitor**
Fixes #4219

changelog: Fix `map_entry` false positive

### CODE REVIEW

File: tests/ui/entry.rs [hunk 1]
```diff
@@ -68,4 +68,11 @@ fn insert_other_if_absent<K: Eq + Hash, V>(m: &mut HashMap<K, V>, k: K, o: K, v:
     }
 }
 
+// should not trigger, because the one uses different HashMap from another one
+fn insert_other<K: Eq + Hash, V>(m: &mut HashMap<K, V>, n: &mut HashMap<K, V>, k: K, v: V) {
+    if !m.contains_key(&k) {
+        n.insert(k, v);
+    }
+}
+
```
Comment (phansch) [inline tests/ui/entry.rs]: Could you also add the following test as described in the issue (where one is mut and the other is not):

```rust
fn insert_other<K: Eq + Hash, V>(m: HashMap<K, V>, n: &mut HashMap<K, V>, k: K, v: V) {
    if !m.contains_key(&k) {
        n.insert(k, v);
    }
}
```

File: clippy_lints/src/entry.rs [hunk 2]
```diff
@@ -140,6 +140,7 @@ impl<'a, 'tcx, 'b> Visitor<'tcx> for InsertVisitor<'a, 'tcx, 'b> {
             if path.ident.name == sym!(insert);
             if get_item_name(self.cx, self.map) == get_item_name(self.cx, &params[0]);
             if SpanlessEq::new(self.cx).eq_expr(self.key, &params[1]);
+            if snippet_opt(self.cx, self.map.span) == snippet_opt(self.cx, params[0].span);
```
Comment (phansch) [inline clippy_lints/src/entry.rs]: Hmm, in general I don't think comparing the span's text is very robust, but I think in this case it should be fine :+1:

**Comment by phansch** [comment]
r=me with the test added :)

File: tests/ui/entry.rs (hunk 1 above)
Comment (JohnTitor) [reply tests/ui/entry.rs]: added the test (and renamed)

**Pull request merged**